Classweights

tdhd opened the pull request:
Picking up on #177 I added class weights to the `Sequential` model.

It works by calling `fit` or `train` with a `class_weight` parameter which must be a dictionary mapping from class labels to class weights. Example taken from the added (MNIST) test case:
```python
high_weight = 100
class_weight = {0:1,1:1,2:1,3:1,4:1,5:1,6:1,7:1,8:1,9:high_weight}
```
The class weights are used to calculate a weight vector to rescale the objective function.

I chose to only add the weight vector to the training functions. Because of that there needs to be a check in the objective functions if the weight vector exists. Looking back at the code right now makes the optional parameter to the objectives look ugly. What do you think? An alternative would be to simply pass `np.ones` to the objective also at test time.

I will clean up the code and add adapted implementations for the remaining objective functions if you want to merge it. At the moment I only added weights to `categorical_crossentropy` for the test case.

Sample output from the test case (left with class weights, right with equal class weights):

```
Classification accuracies on test set:
Digit 0: class_weight = 1 -> 0.941 	 class_weight = 1 -> 0.976
Digit 1: class_weight = 1 -> 0.929 	 class_weight = 1 -> 1.000
Digit 2: class_weight = 1 -> 0.871 	 class_weight = 1 -> 0.862
Digit 3: class_weight = 1 -> 0.766 	 class_weight = 1 -> 0.841
Digit 4: class_weight = 1 -> 0.882 	 class_weight = 1 -> 0.918
Digit 5: class_weight = 1 -> 0.770 	 class_weight = 1 -> 0.885
Digit 6: class_weight = 1 -> 0.908 	 class_weight = 1 -> 0.897
Digit 7: class_weight = 1 -> 0.657 	 class_weight = 1 -> 0.939
Digit 8: class_weight = 1 -> 0.685 	 class_weight = 1 -> 0.798
Digit 9: class_weight = 100 -> 0.989 	 class_weight = 1 -> 0.819
```

tdhd commented:
Do you want to merge this? If yes, I would adapt the implementation of the remaining objectives accordingly and possibly if needed, add an uniform weight vector to the test functions to remove the if conditions in the objectives.

fchollet commented:
I was reviewing it, I made a couple changes. https://github.com/fchollet/keras/tree/tdhd-classweights

My main blocker was having to modify all objective functions. Also I think passing None instead of a unit array would be better, if no weights are provided (it's more general, less brittle). We would have to test the performance impact of that if condition.

tdhd commented:
I checked on the performance impact. The `if` does not have any measurable effect.

The bigger performance issue though is `calculate_class_weights()`. Models trained without explicit class weights (taking `else` branch in function) take on average about 1% less time to train than models with class weights (`if` branch in function). This makes sense because the function is called for every epoch and every batch and calculates an argmax and a mapped dictionary lookup in the `if` branch. This performance effect is driven relatively strongly by the batch size.

The 1% is a relatively noisy estimate. I cannot tell for sure if 1% is the real difference or if it just due to chance 

After a little thought I agree that using `None` instead of a uniform weight vector does make more sense although it inflates the objective functions a little. Additionally, passing the weight vector at evaluation time too (thus making the weight a unit vector) would have an even greater performance impact on model fitting.

I was trying to get the class weights to work with `binary_crossentropy` but without success. Theano was complaining about an dimension error, will check that out again. When I have working examples with all objective functions it would be a good idea to review the options for tuning the performance of `calculate_class_weights`.

In the last commit I synced the PR with your changes.

tdhd commented:
Training with `class_weights` and `binary_crossentropy` is working now. There was a case missing in `calculate_class_weights` where `Y.shape[1] == 1` which would yield an array which is not compatible with `map`.

tdhd commented:
I added the weight vector to `hinge` and `squared_hinge` objectives.

It does not really make sense to add the weight vector to the MSE/MAE objectives. But since the code has to pass the `class_weight` vector to the objective, which is hardcoded in `fit` we somehow have to deal with the weight vector as a parameter.

It would be possible to just hand over a unit weight vector during training, as it is done at the moment for the other objectives. In case the model is instantiated with the MSE/MAE, it should issue a warning when also using the `class_weight` and set it to `None` such that the weight vector becomes a unit weight vector for the objectives. I added the warning in `fit` and in `train`.

tdhd commented:
Fixed the merge conflicts now. I wasn't able to set up a negative test case for the MSE/MAE losses though because of a dimensionality error in theano.

How do you want to proceed?

tdhd commented:
@fchollet Any update on this? :)

fchollet commented:
> I wasn't able to set up a test case for the MSE/MAE losses though because of a dimensionality error in theano.

Did you figure out a solution?

We'll definitely merge this, but I'm still reviewing it. Also, I have been thinking of extending the weighting system to allow per-sample weights, which are useful in many cases (a recent example was Higgs boson decay event detection). What do you think?

tdhd commented:
No I haven't yet. I will have time to look at it this weekend.

Allowing per-sample weights is a good idea, I have never heard about it before though. For that the parametrization of `fit` would have to be changed to allow to pass not only class weights but also per-sample weights.

I will let you know when I have fixed the bug in MSE/MAE.

tdhd commented:
@fchollet, just to clarify: Keras can only fit classification models at the moment, right? Somehow I instantly thought MSE/MAE were to be used for regression models.

Anyway, the training now works with slightly adapted objectives. For example MSE

```python
def mean_squared_error(y_true, y_pred, weight=None):
    if weight is not None:
        return T.sqr(weight*(y_pred.T - y_true.T)).mean()
    else:
        return T.sqr(y_pred.T - y_true.T).mean()
```

where I had to transpose the label vectors. I don't really know why this is solving the input dimension error I received earlier. Do you have any idea? In the test I run, both the weight vector and the training labels are 1D arrays.

fchollet commented on keras/objectives.py:
```diff
@@ -6,29 +6,50 @@
+def binary_crossentropy(y_true, y_pred, weight=None):
     y_pred = T.clip(y_pred, epsilon, 1.0 - epsilon)
-    return T.nnet.binary_crossentropy(y_pred, y_true).mean()
+    bce = T.nnet.binary_crossentropy(y_pred, y_true).mean()
```
Taking the mean makes this quantity a scalar, I believe. It should be a vector (different value for each sample).

fchollet commented:
> @fchollet, just to clarify: Keras can only fit classification models at the moment, right? Somehow I instantly thought MSE/MAE were to be used for regression models.

Both classification and regression use essentially the same workflow. MSE/MAE can also be used in classification models. All losses except crossentropies can be used for both regression and classification.  binary_crossentropy and categorical_crossentropy are only for classification.

> where I had to transpose the label vectors. I don't really know why this is solving the input dimension error I received earlier. Do you have any idea? In the test I run, both the weight vector and the training labels are 1D arrays.

I'm not sure what you mean? What's the shape of the arrays involved?

tdhd commented:
Hey François, today I was getting closer to the dimensionality problem, but still not resolved fully.

The weight vector `w` is created in the case of no class weights as `w = np.ones((Y.shape[0]))` to be, for example `w.shape == (N,)`. Inside the objective functions this same weight vector is then transposed to be `w.shape == (1,N)` though. This explains why the element wise multiplication with `y_pred - y_true` does not work, since they are of shape `(N,1)`.

It is weird though, that this error only happens in the MSE/MAE objectives and not the ones with crossentropy.

fchollet commented:
> The weight vector w is created, in the case of no class weights, as w = np.ones((Y.shape[0])) to be, for example w.shape == (N,). Inside the objective functions this same weight vector is then transposed to be w.shape == (1,N) though. This explains why the element wise multiplication with y_pred - y_true does not work, since they are of shape (N,1).

I think an explanation could be that in the crossentropy objectives, the multiplication happens with a vector of shape `(N,)` (the output of the Theano crossentropy function), which is compatible with the `weight` vector. In MAE/MSE however, the multiplication would happen with a vector of shape `(N, 1)`, so we'd have an incompatibility.

tdhd commented:
I finally solved the dimensionality issue with an explicit reshape of the weight vector in the objectives that required it. In all the objectives, except for categorical crossentropy, the code assumes, that the labels are vectors and not matrices. I think that assumption is ok, since only in multi class problems the labels are one-hot encoded.

In the test cases for (squared) hinge, the network does not seem to learn anything (with or without class weights). I don't know if this is an issue with the network I created for the tests or with the objectives themselves.

In any case, if you want me to supply additional test cases for the remaining objectives, please let me know.

fchollet commented:
> When I get the training working with the hinge losses I think this PR might be finally finished. I can add the instance weights in a different PR. It should be a small change in the weight mapping.

Ok, just let me know when it's ready! : ) Per-sample weight support can definitely be added in a later PR.

tdhd commented:
Ok it's done. All losses can now be scaled with class weights.

The problem weren't the hinge objectives but rather how I set up the test case for them. In the final layer I was using sigmoid activation when the labels were -1 and 1. This if course didn't make any sense as sigmoid never maps to -1. I replaced sigmoid with tanh, which allows the model to learn with the hinge losses.

fchollet commented on keras/models.py:
```diff
@@ -41,6 +41,18 @@ def slice_X(X, start=None, stop=None):
+            y_classes = Y
+        w = np.array(map(lambda x: class_weight[x], y_classes))
+    else:
+        w = np.ones((Y.shape[0]))
```
Why not return `None` instead? Wouldn't it be more efficient (one less element-wise multiplication in the objective function).

tdhd commented:
When setting `w = None` theano somehow manages to map the weight vector inside the losses to a `TensorType(float64, vector)`. For that reason the `None` check does not work but during runtime the computation results in an error:

```
ValueError: expected an ndarray, not None
Apply node that caused the error: Elemwise{mul,no_inplace}(<TensorType(float64, vector)>, Elemwise{neg,no_inplace}.0)
Inputs types: [TensorType(float64, vector), TensorType(float64, vector)]
Inputs shapes: ['No shapes', (128,)]
Inputs strides: ['No strides', (8,)]
Inputs values: [None, 'not shown']
```

So although the contents of the vector are `None` the `if` check does not work. I tried checking with `T.isnan` which removes `weight` from the compuational graph completely.

The weight vector only defaults to `None` in `_predict`, `_test` and `_test_with_acc` where the weight vector is not passed as an argument at all.

I suggest we keep the implementation with `np.ones`. In a later stage, if it is necessary from a performance viewpoint, the issue can still be addressed. What do you think?

fchollet commented:
Works for me. 

Merging this now. Thank you for putting this feature together!

harsimrat-eyeem commented:
@fchollet @tdhd 

Hi,

The is an issue with using **class_weights + categorical_crossentropy + multiple labels**. i.e k-hot vectors. with k>1. 

In : [models.py] (https://github.com/fchollet/keras/blob/master/keras/models.py#L88)

For the case of [ k-hot binary vector as labels ] ( https://github.com/fchollet/keras/blob/master/keras/models.py#L97 ) keeps the weight for the class index given by sorting y along label axis and using weight for the class where first hot bit is encountered and **not** rest of the hot bits. This would be okay for k=1, but doesn't generalise. 

sample case below. 

```
import numpy as np

y = p.random.binomial(1,0.5,1000).reshape(10,100)
class_weight = {idx-1:1/(idx+0.0) for idx in range(1,101)}

def standardize_weights(y, sample_weight=None, class_weight=None):
    if sample_weight is not None:
        return standardize_y(sample_weight)
    elif isinstance(class_weight, dict):
        if len(y.shape) > 3:
            raise Exception('class_weight not supported for 4+ dimensional targets.')
        yshape = y.shape
        # for time-distributed data, collapse time and sample
        y = np.reshape(y, (-1, yshape[-1]))
        if y.shape[1] > 1:
            y_classes = y.argmax(axis=1)
        elif y.shape[1] == 1:
            y_classes = np.reshape(y, y.shape[0])
        else:
            y_classes = y
        class_weights = np.asarray([class_weight[cls] for cls in y_classes])
        return np.reshape(class_weights, yshape[:-1] + (1,))  # uncollapse initial dimensions
    else:
        return np.ones(y.shape[:-1] + (1,))

weights = standardize_weights(y, class_weight= class_weight)

```
results in samplesX1 weight vector and **not** samplesXnb_classes since it ignores all the classes except the indices where it found the first hot bit. This ends up being the sample weight in the outer loop of the loss function ( over the samples ) but not in the inner loop ( over the classes ).

[[ 0.33333333]
 [ 1.        ]
 [ 0.33333333]
 [ 0.25      ]
 [ 1.        ]
 [ 0.33333333]
 [ 1.        ]
 [ 1.        ]
 [ 1.        ]
 [ 0.5       ]]
